control/velocity_controller_lqr/velocity_controller_lqr/velocity_controller_lqr_lib.py

```diff
@@ -0,0 +1,289 @@
+    ) -> np.array:
+        """Calculates the 3x3 coriolis matrix.
+
+        Parameters:
+            pitch_rate: float: pitch rate in rad/s
+            yaw_rate: float: yaw rate in rad/s
+            sway: float: sway velocity in m/s
+            heave: float: heave velocity in m/s
+
+        Returns:
+            C: np.array: 3x3 Coriolis Matrix
+
+        """
+        return np.array(
+            [
+                [0.2, -30 * sway * 0.01, -30 * heave * 0.01],
+                [30 * sway * 0.01, 0, 1.629 * pitch_rate],
+                [30 * heave * 0.01, 1.769 * yaw_rate, 0],
+            ]
+        )
+
+    def calculate_lqr_u(
+        self, coriolis_matrix: np.array, states: np.array, guidance_values: np.array
+    ) -> np.array:
+        """Calculates the control input using the control library in python.
+
+        Parameters:
+            C: np.array: 3x3 Coriolis Matrix
+            states: np.array: 6x1 state vector
+            guidance_values: np.array: 6x1 guidance vector
+
+        Returns:
+            u: np.array: 3x1 control input
+
+        """
+        system_matrix = self.inertia_matrix_inv @ coriolis_matrix
+        input_matrix = self.inertia_matrix_inv
+
+        # Augment the A and B matrices for integrators for surge, pitch, and yaw
+        self.augmented_system_matrix = np.block(
+            [[system_matrix, np.zeros((3, 3))], [-np.eye(3), np.zeros((3, 3))]]
+        )  # Integrators added for surge, pitch, and yaw
+        self.augmented_input_matrix = np.block(
+            [[input_matrix], [np.zeros((3, 3))]]
+        )  # Control input does not affect integrators directly
+
+        # CT LQR controller from control library python
+        lqr_gain, ricatti_s, ricatti_e = ct.lqr(
+            self.augmented_system_matrix,
+            self.augmented_input_matrix,
+            self.state_weight_matrix,
+            self.input_weight_matrix,
+        )
+
+        # Calculate the control input
+        surge_error = (
+            guidance_values[0] - states[0]
+        )  # Surge error no need for angle wrapping
+        pitch_error = self.ssa(
+            guidance_values[1] - states[1]
+        )  # Apply ssa to pitch error
+        yaw_error = self.ssa(guidance_values[2] - states[2])  # Apply ssa to yaw error
+
+        # Update the integrators for surge, pitch, and yaw
+        self.integral_error_surge = self.anti_windup(
+            self.i_surge, surge_error, self.integral_error_surge, self.surge_windup
+        )
+        self.integral_error_pitch = self.anti_windup(
+            self.i_pitch, pitch_error, self.integral_error_pitch, self.pitch_windup
+        )
+        self.integral_error_yaw = self.anti_windup(
+            self.i_yaw, yaw_error, self.integral_error_yaw, self.yaw_windup
+        )
+
+        state_error = np.array(
+            [
+                -surge_error,
+                -pitch_error,
+                -yaw_error,
```

Comment:
... these can be positive

Reply:
if i flip, then the integral values become negative. Pick your poison.